File: ios/CircleBoxSDK/Sources/CircleBoxSDK/CircleBoxRingBuffer.swift
import Foundation

/// Fixed-capacity, thread-safe circular buffer.
///
/// When full, appending a new element overwrites the oldest element.
final class CircleBoxRingBuffer<Element> {
    private let queue = DispatchQueue(label: "com.circlebox.sdk.ringbuffer")
    private var storage: [Element?]
    private var writeIndex: Int = 0
    private var storedCount: Int = 0
    private var sequence: Int64 = 0

    init(capacity: Int) {
        precondition(capacity > 0)
        self.storage = Array(repeating: nil, count: capacity)
    }

    var capacity: Int {
        storage.count
    }

    var count: Int {
        queue.sync { storedCount }
    }

    func append(_ build: (Int64) -> Element) {
        queue.sync {
            let next = sequence
            sequence += 1
            storage[writeIndex] = build(next)
            writeIndex = (writeIndex + 1) % storage.count
            storedCount = min(storedCount + 1, storage.count)
        }
    }

    func snapshot() -> [Element] {
        queue.sync {
            guard storedCount > 0 else { return [] }
            var result: [Element] = []
            result.reserveCapacity(storedCount)
            // Once the buffer reaches capacity, writeIndex always points to the oldest item.
            let start = storedCount == storage.count ? writeIndex : 0
            for offset in 0..<storedCount {
                let index = (start + offset) % storage.count
                if let value = storage[index] {
                    result.append(value)
                }
            }
            return result
        }
    }

    func clear() {
        queue.sync {
            storage = Array(repeating: nil, count: storage.count)
            writeIndex = 0
            storedCount = 0
        }
    }
}

File: android/circlebox-sdk/src/main/kotlin/com/circlebox/sdk/CircleBoxRingBuffer.kt
package com.circlebox.sdk

import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Fixed-capacity circular buffer protected by a [Mutex].
 *
 * Writes are synchronous on purpose so crash-time snapshots always observe committed events.
 */
internal class CircleBoxRingBuffer<T>(capacity: Int) {
    private val capacity = capacity.coerceAtLeast(1)
    private val mutex = Mutex()
    private val storage = arrayOfNulls<Any>(this.capacity)
    private var writeIndex = 0
    private var storedCount = 0
    private var sequence: Long = 0

    fun append(factory: (Long) -> T) {
        runBlocking {
            mutex.withLock {
                val next = sequence++
                storage[writeIndex] = factory(next)
                writeIndex = (writeIndex + 1) % this@CircleBoxRingBuffer.capacity
                storedCount = minOf(storedCount + 1, this@CircleBoxRingBuffer.capacity)
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun snapshot(): List<T> = runBlocking {
        mutex.withLock {
            if (storedCount == 0) {
                return@withLock emptyList()
            }

            val result = ArrayList<T>(storedCount)
            // Once full, writeIndex points at the oldest item in the buffer.
            val start = if (storedCount == capacity) writeIndex else 0
            for (offset in 0 until storedCount) {
                val index = (start + offset) % capacity
                val value = storage[index] as T?
                if (value != null) {
                    result.add(value)
                }
            }
            result
        }
    }

    fun count(): Int = runBlocking {
        mutex.withLock { storedCount }
    }
}
